src/widgets/concept-widget.tsx

```diff
@@ -0,0 +1,148 @@
+import {
+  VDomModel, VDomRenderer
+} from '@jupyterlab/apputils';
+import {INotebookTracker} from '@jupyterlab/notebook';
+
+import * as React from 'react';
+
+import {AllOfUsConfig} from '../config';
+import {Concept} from '../generated';
+import {ConceptsService} from '../services/concepts.service';
+
+import {Observable} from 'rxjs/Rx';
+
+export enum LoadingStatus {
+  UNSET,
+  LOADING,
+  LOADED
+}
+
+export class ConceptsWidgetModel extends VDomModel {
+
+  public initialized: boolean;
+  public loadingStatus = LoadingStatus.UNSET;
+  public query = '';
+  public concepts: Array<Concept>;
+  public selectedConcept: Concept;
+}
+
+export class ConceptsWidgetRenderer extends VDomRenderer<ConceptsWidgetModel> {
+
+
+  constructor(configObservable: Observable<AllOfUsConfig>,
+              private conceptsService: ConceptsService,
+              private notebooks: INotebookTracker) {
+    super();
+
+    this.id = 'allofus-concepts';
+    this.title.label = 'Concepts';
+    this.title.closable = true;
+    configObservable.subscribe((config) => {
+      if (!this.model.initialized) {
+        this.model.initialized = true;
+        this.update();
+      }
+    });
+  }
+
+  protected render(): React.ReactElement<any> {
+    const handleQueryChange = (event) => {
+      this.model.query = event.target.value;
+      this.update();
+    };
+
+    const searchConcepts = () => {
+      this.model.loadingStatus = LoadingStatus.LOADING;
+      this.update();
+      this.conceptsService.searchConcepts({query: this.model.query})
+          .then((concepts) => {
+            this.model.concepts = concepts;
+            this.model.loadingStatus = LoadingStatus.LOADED;
+            this.update();
+          });
+    };
+    return (
+        <div>
+          <input value={this.model.query} onChange={handleQueryChange}
+                 disabled={!this.model.initialized}/>
+          <br/>
+          <input value='Search' type='button' onClick={searchConcepts}
+                 disabled={!this.model.initialized || !this.model.query}/>
+          <br/>
+          {this.loadingImage()}
```

Comment:
Note: some other approaches to composition are listed here: https://medium.com/dailyjs/techniques-for-decomposing-react-components-e8a1081ef5da

But the approach you have now definitely makes sense if we're not really doing data binding here and using the React rerendering stuff.